cmd/allocator/main.go

```diff
@@ -0,0 +1,165 @@
+// Copyright 2019 Google LLC All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package main
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"strings"
+
+	allocationv1alpha1 "agones.dev/agones/pkg/apis/allocation/v1alpha1"
+	"agones.dev/agones/pkg/client/clientset/versioned"
+	"agones.dev/agones/pkg/util/runtime"
+	k8serror "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/rest"
+)
+
+var (
+	logger = runtime.NewLoggerWithSource("main")
+)
+
+const (
+	certDir = "/home/allocator/client-certs/"
+	tlsDir  = "/home/allocator/tls/"
+	port    = "8443"
+)
+
+// A handler for the web server
+type handler func(w http.ResponseWriter, r *http.Request)
+
+func main() {
+	agonesClient, err := getAgonesClient()
+	if err != nil {
+		logger.WithError(err).Fatal("could not create agones client")
+	}
+
+	h := httpHandler{
+		agonesClient: agonesClient,
+		namespace:    os.Getenv("NAMESPACE"),
+	}
+
+	// TODO: add liveness probe
+	http.HandleFunc("/v1alpha1/gameserverallocation", h.postOnly(h.allocateHandler))
+
+	caCertPool, err := getCaCertPool(certDir)
+	if err != nil {
+		logger.WithError(err).Fatal("could not get CA certs")
+	}
+
+	cfg := &tls.Config{
+		ClientAuth: tls.RequireAndVerifyClientCert,
+		ClientCAs:  caCertPool,
+	}
+	srv := &http.Server{
+		Addr:      ":" + port,
+		TLSConfig: cfg,
+	}
+
+	err = srv.ListenAndServeTLS(tlsDir+"tls.crt", tlsDir+"tls.key")
+	logger.WithError(err).Fatal("allocation service crashed")
+}
+
+// Set up our client which we will use to call the API
+func getAgonesClient() (*versioned.Clientset, error) {
+	// Create the in-cluster config
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, errors.New("Could not create in cluster config")
+	}
+
+	// Access to the Agones resources through the Agones Clientset
+	agonesClient, err := versioned.NewForConfig(config)
+	if err != nil {
+		return nil, errors.New("Could not create the agones api clientset")
+	}
+
+	return agonesClient, nil
+}
+
+func getCaCertPool(path string) (*x509.CertPool, error) {
+	// Add all certificates under client-certs path because there could be multiple clusters
+	// and all client certs should be added.
```

Comment:
Normally this should be one CA certificate regardless of # of clusters, clients will use leaf certificates signed by this CA, so you technically only need one PEM file with all CA certificate bundle, but not all the client certs.

Reply:
IOW - we need make sure we have the certificates example set up correctly, because people will be cargo-culting this a lot.

Reply:
This acts as a revocation list for certificates as well. If a secret is compromised, its CA should be revoked without impacting other clients calling allocation service. Having this pattern is valuable for match making and not for cluster to cluster calls. If a cluster secret is compromised, then all secrets to talk to another cluster is compromised. However, matchmaker is an independent entity and this solution helps securing system from matchmakers bad actors.